### PR TITLE
Add mutex to V3Error class

### PR DESCRIPTION
**Pull request by kamilrakoczy**
Related to: https://github.com/verilator/verilator/pull/3608

In this PR, I'm adding mutex V3Error class to make printing error messages more thread-safe. I is not ideal solution of the problem, but it should ensure that different error messages are not mixed together.

~~I'm also changing configure to enable thread-safety checks for clang and define ``VL_THREADED`` as it is required to enable mutex attributes.~~ I'm using ``VerilatedMutex``/``VerilatedLockGuard`` as they have set attributes required for thread-safe checks.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -409,6 +409,8 @@ _MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-Qunused-arguments)
 _MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-faligned-new)
 _MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-Wno-unused-parameter)
 _MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-Wno-shadow)
+_MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-Wthread-safety)
+_MY_CXX_CHECK_OPT(CFG_CXXFLAGS_SRC,-DVL_THREADED)
```
Comment (wsnyder) [inline configure.ac:413]: I think that configure/make changes should be a separate patch.  But
1. We should not be setting VL_THREADED based on a compiler switch.  I think all compilers we support have C++11 now, so we can always be setting VL_THREADED when compiling Verilator.
2. Thread-safety is already turned on in the WEXTRA flags, which is where it should be.

File: src/V3Error.h [hunk 2]
```diff
@@ -276,33 +283,41 @@ class V3Error final {
...
+    static V3ErrorCode errorCode() {
+        const VerilatedLockGuard lg{s_mutex};
+        return s_errorCode;
+    }
+    static void errorExitCb(ErrorExitCb cb) VL_REQUIRES(s_mutex) { s_errorExitCb = cb; }
 
     // When printing an error/warning, print prefix for multiline message
-    static string warnMore();
+    static string warnMore() VL_MT_SAFE;
+    static string warnMoreNoLock() VL_REQUIRES(s_mutex) VL_MT_SAFE;
```
Comment (wsnyder) [inline src/V3Error.h:320]: I think the mix of lock and no-lock functions is hard to maintain.  I suggest instead make a separate class with all protected or private members, which assumes that the lock has been obtained and those functions are only called from that class, or from this class which first obtains the lock.

**Review by wsnyder — CHANGES_REQUESTED** [review]
#3686 needed first.

File: src/V3Error.cpp [hunk 3]
```diff
@@ -89,7 +92,7 @@ string V3Error::lineStr(const char* filename, int lineno) {
     return out.str();
 }
 
-void V3Error::incErrors() {
+void V3Error::incErrors() VL_MT_SAFE {
     s_errCount++;
```
Comment (wsnyder) [inline src/V3Error.cpp:96]: ++ isn't safe

File: src/V3Error.h [hunk 4]
```diff
@@ -18,7 +18,8 @@
 #define VERILATOR_V3ERROR_H_
 
 #include "config_build.h"
-#include "verilatedos.h"
+
+#include "verilated.h"
```
Comment (wsnyder) [inline src/V3Error.h:22]: Include differently, discussed in other pull.

File: src/V3Error.h [hunk 5]
```diff
@@ -246,33 +247,52 @@ inline std::ostream& operator<<(std::ostream& os, const V3ErrorCode& rhs) {
     return os << rhs.ascii();
 }
 
-// ######################################################################
-
-class V3Error final {
-    // Base class for any object that wants debugging and error reporting
-
+class V3ErrorGuarded VL_NOT_FINAL {
```
Comment (wsnyder) [inline src/V3Error.h:250]: Need comments above on what this class is for.

File: src/V3Error.h [hunk 6]
```diff
@@ -246,33 +247,52 @@ inline std::ostream& operator<<(std::ostream& os, const V3ErrorCode& rhs) {
...
+
+    static string msgPrefixNoLock() VL_REQUIRES(s_mutex) VL_MT_SAFE;  // returns %Error/%Warn
+    static string warnMoreNoLock() VL_REQUIRES(s_mutex) VL_MT_SAFE;
+    static void errorExitCbNoLock(ErrorExitCb cb) VL_REQUIRES(s_mutex) { s_errorExitCb = cb; }
+    static bool isError(V3ErrorCode code, bool supp) VL_REQUIRES(s_mutex) VL_MT_SAFE;
+    static void vlAbortOrExit() VL_REQUIRES(s_mutex) VL_MT_SAFE;
+};
+
+// ######################################################################
+class V3Error final : protected V3ErrorGuarded {
```
Comment (wsnyder) [inline src/V3Error.h:285]:
```suggestion

class V3Error final : protected V3ErrorGuarded {
```

An is-a relationship might not be ideal here, as it makes it easy to mistakenly call a unguarded function.  What does a has-a relationship with guarded look like instead?

**Comment by kamilrakoczy** [comment]
Requires: https://github.com/verilator/verilator/pull/3822

File: src/V3Error.h [hunk 7]
```diff
@@ -156,8 +157,9 @@ class V3ErrorCode final {
     V3ErrorCode()
         : m_e{EC_MIN} {}
     // cppcheck-suppress noExplicitConstructor
-    constexpr V3ErrorCode(en _e)
-        : m_e{_e} {}
+    constexpr V3ErrorCode(en _e) VL_PURE : m_e{_e} {}
```
[the remark this answers is not in the document]
Comment (mglb) [reply src/V3Error.h:160]: FYI: seems like the CI compiles Verilator using  C++ version newer than 11, where this assumption is not true. Maybe it would be good to test with C++11?
For an example of non-pure constexpr method (invalid in C++11, valid since C++14) see the code two lines below:
```c++
constexpr void operator=(const V3ErrorCode& rhs) VL_PURE { m_e = rhs.m_e; }
```

File: src/V3Error.cpp [hunk 8]
```diff
@@ -175,27 +173,34 @@ void V3Error::vlAbort() {
...
-string V3Error::warnMore() { return string(msgPrefix().size(), ' '); }
+string V3Error::warnMore() VL_MT_SAFE_EXCLUDES(singleton().m_mutex) {
+    const VerilatedLockGuard guard{singleton().m_mutex};
+    return singleton().warnMore();
+}
 
 // cppcheck-has-bug-suppress constParameter
-void V3Error::v3errorEnd(std::ostringstream& sstr, const string& extra) {
+void V3Error::v3errorEnd(std::ostringstream& sstr, const string& extra)
+    VL_MT_SAFE_EXCLUDES(singleton().m_mutex) {
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/V3Error.cpp]: I've added ``VL_MT_SAFE_STABLE_TREE``.

**Comment by kamilrakoczy** [comment]
Requires: https://github.com/verilator/verilator/pull/3838

In this update I've changed when we are starting locking ``V3ErrorGuarded`` mutex. Now this mutex is getting locked as soon as we are entering ``v3warnCode`` macro.

Unfortunately due to current clang limitations, we can't use ``VerilatedLockGuard`` inside macro ``(..., ...)`` operator as it doesn't properly detects it (clang-truck version already properly detects it, but clang-15 still not). This forced me to use ``,`` operator to properly lock and unlock mutex there: https://github.com/verilator/verilator/pull/3680/files#diff-0d54df7e79ae6321d2179f98eafa7b415ba964ee2e42160ab9737f70402341ccR511.

I've also noticed, that due to some bug in ``gcc``, we can't directly use ``s().v3errorEnd`` in global ``v3errorEnd`` as it causes some internal error inside gcc when it prints stacktrace (see: https://github.com/antmicro/verilator-1/actions/runs/3799095496/jobs/6461396979#step:6:180). As a workaround, I've created ``v3errorEndGuardedCall`` function inside ``V3Error`` that requires locked mutex and just calls ``s().v3errorEnd``.

File: src/V3Error.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/V3Error.h:161]: As far I know, currently libclang doesn't provide a way to check if function is ``contexpr``. We could check tokens, but it doesn't follow macros, typedefs etc. 

(Some old answer from stackoverflow: https://stackoverflow.com/questions/60627345/python-libclang-how-to-distinguish-const-vs-constexpr-types)

File: include/verilatedos.h [hunk 9]
```diff
@@ -161,6 +161,13 @@
 # define VL_MT_SAFE
 #endif
 // Comment tag that function is threadsafe, only
+// when other theads doesn't change AST tree
+#if defined(__clang__)
+# define VL_MT_SAFE_STABLE_TREE __attribute__((annotate("MT_SAFE")))
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply include/verilatedos.h:166]: Just to make sure I understood your comment correctly: I should add ``VL_MT_SAFE_STABLE`` definition (as I can't see ``stable`` defined already) and add check that ``VL_MT_SAFE_STABLE`` doesn't call ``VL_MT_SAFE_STABLE_TREE``? If yes, when node should be annotated with ``VL_MT_SAFE_STABLE``?

File: test_regress/t/t_flag_topmodule_bad.out [hunk 10]
```diff
@@ -1,7 +1,5 @@
 %Warning-MULTITOP: t/t_flag_topmodule.v:15:8: Multiple top level modules
                                             : ... Suggest see manual; fix the duplicates, or use --top-module to select top.
-                   ... For warning description see https://verilator.org/warn/MULTITOP?v=latest
-                   ... Use "/* verilator lint_off MULTITOP */" and lint_on around source to disable this message.
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply test_regress/t/t_flag_topmodule_bad.out:4]: I've changed this on purpose, but I can think about different solution if you prefer printing this as a first message.
Let me first explain reasoning about this change:

Before my change, we were printing error message and than adding some extra information by directly printing on ``cerr``: https://github.com/verilator/verilator/blob/master/src/V3LinkLevel.cpp#L67, this is ok if we assume, that ``m_errorCode`` won't change between ``v3warn`` call and printing extra information (as when we are printing extra information we are using ``warnMore`` function that calculates how many extra spaces print to align with previous error message).

This isn't true anymore after my change,V3Error mutex is locked when we are inside ``v3warn`` macro, but after this macro, some other thread can lock this mutex and change ``m_errorCode``. To fix this issue, I've moved the code that is responsible of generating this extra information to lambda that is executed inside ``v3warn`` context (so when we are still holding V3Error mutex).

This isn't big concern for now (yet) as we are currently only aiming to parallel emit stage, but in the future we would need to remove all usages of ``warnMore`` that aren't called from macro that is locking V3Error mutex.

To revert this change I can add some extra option that allows to add extra message after printing ``lint_off`` warning or
simply revert my change and leave fixing this when we will parallel ``V3LinkLevel``.

File: src/verilog.y [hunk 11]
```diff
@@ -343,7 +343,7 @@ static void ERRSVKWD(FileLine* fileline, const string& tokname) {
     fileline->v3error(
         std::string{"Unexpected '"} + tokname + "': '" + tokname
         + "' is a SystemVerilog keyword misused as an identifier."
-        + (!toldonce++ ? "\n" + V3Error::warnMore()
+        + (!toldonce++ ? "\n" + V3Error::s().warnMore()
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/verilog.y:346]: I've explained it in more details in my answer above.

There are still some places in the code where we are calling ``V3Error::warnMore()`` outside of ``v3warn`` macros, so I need to have function that doesn't require already locked V3Error mutex.

If you are OK with this, I can create new function that would be current ``V3Error::warnMore()`` (so it would excludes V3Error mutex) and it could be used outside ``v3warn`` macros, then I could create ``V3Error::warnMore()`` function that requires V3Error mutex that would simply call ``s().warnMore()``.

File: src/V3SchedAcyclic.cpp [hunk 12]
```diff
@@ -99,13 +99,13 @@ class Graph final : public V3Graph {
         // TODO: 'typeName' is an internal thing. This should be more human readable.
         if (LogicVertex* const lvtxp = dynamic_cast<LogicVertex*>(vtxp)) {
             AstNode* const logicp = lvtxp->logicp();
-            std::cerr << logicp->fileline()->warnOther()
+            std::cerr << logicp->fileline()->warnOtherGuarded()
```
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/V3SchedAcyclic.cpp:102]: Maybe function name is a little bit misleading here, both functions ``warnOther`` and ``warnOtherGuarded`` are guarded, but the difference is what function locks the V3Error mutex.

In ``warnOther`` variant, the mutex is locked by ``v3warn`` macro and in ``warnOtherGuarded`` variant, this function itself locks V3Error mutex. I've done this to allow usage of ``warnOther`` outside of ``v3warn`` macro (``warnOther`` function underneath uses ``V3Error::warnMore``: https://github.com/verilator/verilator/blob/master/src/V3FileLine.cpp#L412).

Flipping the name of function would result a lot of changes in the code, every place where we are using ``fileline()->warnOther()`` in the code inside ``v3warn`` macro would need to be adapted. On the other hand there are only 2 cases where it is used outside of ``v3warn`` macro.

File: src/verilog.y (hunk 11 above)
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/verilog.y]: I've added ``warnMoreGuarded`` function that excludes V3Error mutex with annotation that this function isn't MT_SAFE, so when it will be used in MT_SAFE context, ``clang_check_annotations`` script will warn about this,

Now ``V3Error::warnMore`` simply calls ``s().warnMore()`` (and requires V3Error mutex).

File: src/V3SchedAcyclic.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply src/V3SchedAcyclic.cpp:102]: I've added ``VL_MT_UNSAFE`` annotation to ``warnOtherGuarded`` function, so when it will be used in ``MT_SAFE`` context, ``clang_check_annotations`` script will warn about this.

File: include/verilatedos.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply include/verilatedos.h:166]: My comment was confusing, I meant do this

     # define VL_MT_SAFE_STABLE_TREE __attribute__((annotate("MT_SAFE_STABLE_TREE")))

Then update the checker that MT_SAFE_STABLE_TREE can call other MT_SAFE_STABLE_TREE or MT_SAFE or PURE (i.e. anything marked). But MT_SAFE cannot call MT_SAFE_STABLE_TREE as that is "less" safe. The idea is from a safeness standpoint we enforce

     MT_SAFE_STABLE_TREE < MT_SAFE < PURE

File: test_regress/t/t_flag_topmodule_bad.out (hunk 10 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply test_regress/t/t_flag_topmodule_bad.out:4]: I see your point, but I want the message printed after the first warning/error always to help the user.  However you want to accomplish this is fine.

One hack I don't love but might fix the problem is to simply handle it with post processing that is have the warnMore insert the magic text "__WARNMORE__" then post process to replace text appropriately.

File: src/verilog.y (hunk 11 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply src/verilog.y]: See my note above about magic postprocessing, I think that is a cleaner solution (to the V3Errror users anyways).

File: src/V3SchedAcyclic.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (wsnyder) [reply src/V3SchedAcyclic.cpp:102]: Maybe call this warnOtherStandalone then?

File: include/verilatedos.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply include/verilatedos.h:166]: The main goal of this PR is to make ``v3errorEnd`` safe to call from MT-safe context (to allow parallelization). I believe that this check (on ``v3errorEnd`` function) would mean that no part of parallel code can call ``v3errorEnd`` or whole speed-up from parallelization would be gone.

If I'm not wrong, the only part that makes ``v3errorEnd`` require stable tree are functions inside ``V3ERROR_NO_GLOBAL_`` ifndef (https://github.com/verilator/verilator/blob/master/src/V3Error.cpp#L279).
They are only called when error is FATAL and debug or dump tree flags are set. 

What to you think about not calling this function if we are running verilation with multiple threads and instead print some information message that it is only supported on single thread verilation? This would allow to annotate ``v3errorEnd`` with just ``MT_SAFE``.

[the remark this answers is not in the document]
Comment (wsnyder) [reply include/verilatedos.h:166]: I think we definitely do want to dump when multithreaded.  I'm expecting at some point that will be the main operation mode and getting these dumps is critical to debug.

It makes sense that most algorithms will want to have routines that execute in parallel, and inside that parallel code they have UASSERTs.  We don't want to have to single thread those UASSERTs just in case they fire.  How in general will we be doing thread cleanup on fatals?  That answer might suggest how we handle v3fatal.

[the remark this answers is not in the document]
Comment (kamilrakoczy) [reply include/verilatedos.h:166]: Sure, so I think it would be the best to use ``Cooperative Interruption`` concept from C++20 to handle this (https://en.cppreference.com/w/cpp/thread/stop_source / https://www.modernescpp.com/index.php/cooperative-interruption-of-a-thread-in-c-20).

In this concept we can emit request that other threads can check (but this needs to be done manually) in appropriate place and if it is set, stop executing the job.

When we will encourage fatal error that requires dumping the tree, we would emit such signal, wait for other threads to stop execution (by entering function that checks for stop request, so generally any function that requires stable tree, or completes its job).
When all threads will complete its job, we would continue main thread and dump tree.

What do you think about this solution?

As this is mostly used by developers, maybe we can require C++20 for this feature?

**Comment by wsnyder** [comment]
I think the stop request is a good approach.

Debugging is not just for developers, it is the key feedback from users, so I think any use of multithreading needs this.

I'm fine though requiring C++20 for Verilator multithreading, given we need it for coroutines too.

**Comment by kamilrakoczy** [comment]
Finally, I decided not to use ``std::stop_source`` from C++20, as it can't be easily withdrawn (it expects that when stop request is made, threads that will receive it will stop execution). Instead I've used condition variables to synchronize between exclusive access and parallel processing.

I'm not sure how we want to proceed with merging this. This change requires ThreadPool that I cherry-picked from https://github.com/verilator/verilator/pull/3608 (along with parallelization of emit stage, but this isn't necessary). Should I split it to only include ThreadPool and options or we want to do this differently?

**Review by wsnyder — COMMENTED** [review]
I agree with your idea on splitting this up a bit as this is getting hard to review.
I think the first thing you can peel off is the parallelization options, that can be merged even though nothing will run in parallel yet.

Then add the threadpool?  Then the VL_SAFE attributes?  I'm not sure what's best to do next.

**Comment by kamilrakoczy** [comment]
I think this should be the next PR to merge as some functions from annotations PRs requires ``v3Error`` to be MT-safe.

I've added ``lockCheckStopRequest`` function to ``VerilatorMutex``/ ``VerilatorLockGuard`` classes that can periodically check if stop request was send by another thread while trying to lock the mutex. I think better place for this function would be in ``V3ThreadPool``, but due to limitations in clang thread-analysis (it reports that mutex may not be locked even when the only way to return from function is after locking the mutex) it needs to be in this classes.

**Comment by wsnyder** [comment]
In V3Graph:
```
    void v3errorEnd(std::ostringstream& str) const VL_REQUIRES(V3Error::s().m_mutex);
    void v3errorEndFatal(std::ostringstream& str) const VL_REQUIRES(V3Error::s().m_mutex);
```

That custom outputters need what should otherwise be a private mutex seems ugly.  I was thinking about this and maybe we can make all the errorEnd lock free.  What if we pass the custom info that is currently in a static as part of a custom stream object, like this:

```
#include <iostream>
#include <sstream>
#include <string>

class V3ErrorStream;
void v3errorEnd(V3ErrorStream& es) {
//    es.emit();
}

class V3ErrorStream : public std::ostringstream {
    std::string m_additionalInfoLikeFileline;
public:
    V3ErrorStream(const std::string& additionalInfoLikeFileline) :
        m_additionalInfoLikeFileline{additionalInfoLikeFileline} {}
    ~V3ErrorStream() {
        std::cout << "Msg: " << m_additionalInfoLikeFileline << ": " << this->str() << std::endl;
        v3errorEnd(*this);
    }
};

class V3Error {
public:
    static V3ErrorStream v3errorPrep(bool ignore) {
        return V3ErrorStream{"filelineetc"};
    }
};

// Note no longer calling v3errorEnd like this v3errorEnd((V3Error::v3errorPrep(false)<<msg));
#define v3error(msg) do { V3Error::v3errorPrep(false)<<msg; } while (0)

int main() {
    v3error("Testing");
}
```

Thoughts?

**Comment by kamilrakoczy** [comment]
We can't use ``do {} while(0)`` construct and skip ``(...,...)`` as ``v3warn`` is sometimes used on objects (https://github.com/verilator/verilator/blob/master/src/V3EmitCFunc.cpp#L265) and needs to expanded to ``v3errorEnd``/``v3errorEndFatal`` for custom outputers, but using custom stream class to store static variables should be possible.

**Comment by kamilrakoczy** [comment]
Actually even custom stream class requires a lot of changes in verilator code, mostly because of usage of ``warnMore``/``warnOther`` methods that assumes it will return ``string``.

This methods requires current error code and when we will move static ``errorCode`` out from ``V3Error`` to ``V3ErrorStream`` we can't return exact values in this methods.
I've tried to return placeholder class ``warnMore_t`` that later would be replaced with proper string.
It works in most cases, but it breaks usage of terany operator with ``warnMore`` (e.g. https://github.com/verilator/verilator/blob/master/src/verilog.y#L360) as terany actually evaluates operand (and requires both sides to have same type).
We also can't use string placeholders (e.g. ``__WARNMORE__``) as this function sometimes depends also on object (e.g. in case of ``FileLine``).

The lock that is required in custom outputters is automatically locked by ``v3warnCode`` macro and ``VL_REQUIRES(V3Error::s().m_mutex)`` in ``V3Graph::v3errorEnd`` is only information to clang thread-analysis that we already have locked mutex at this point (to warn user when he would use methods from ``V3Error`` class instead of ``V3ErrorGuarded``). User doesn't need to manually lock this mutex in custom outputters.

Maybe we could leave this custom stream class refactor to another PR? @wsnyder what do you think?

**Comment by wsnyder** [comment]
Good points.  I was presuming any stream work would be a separate PR, but probably before this if we did it.  Just was throwing the idea out.  Let me review this as it is again....

File: src/V3LinkLevel.cpp [hunk 13]
```diff
@@ -59,15 +59,23 @@ void V3LinkLevel::modSortByLevel() {
...
+            auto warnTopModules = [](std::string warnMore, ModVec tops)
+                                      VL_REQUIRES(V3Error::s().m_mutex) -> std::string {
+                std::stringstream ss;
+                for (AstNode* alsop : tops) {
+                    ss << warnMore << "... Top module " << alsop->prettyNameQ() << endl
+                       << alsop->warnContextSecondary();
+                }
+                return ss.str();
+            };
+
```
Comment (wsnyder) [inline src/V3LinkLevel.cpp:71]: I don't think warnContextSecondary should need a lock (or at least not the V3Error lock), so you should be able to avoid a lambda and just execute this code here and print ss.str() directly below.

File: src/V3Error.cpp [hunk 14]
```diff
@@ -219,71 +160,80 @@ void V3Error::v3errorEnd(std::ostringstream& sstr, const string& extra) {
...
+                            v3Global.rootp()->dumpTreeFile(
+                                v3Global.debugFilename("final.tree", 990));
+                        }
+                        if (debug()) {
+                            execErrorExitCb();
+                            V3Stats::statsFinalAll(v3Global.rootp());
+                            V3Stats::statsReport();
+                        }
+                        // Abort in exclusive access to make sure other threads
+                        // doesn't change error code
```
Comment (wsnyder) [inline src/V3Error.cpp:234]:
```suggestion
                        // don't change error code
```

File: src/V3Error.h [hunk 15]
```diff
@@ -250,71 +255,219 @@ inline std::ostream& operator<<(std::ostream& os, const V3ErrorCode& rhs) {
...
+    // generate whole error message inside v3warn macros and it needs to be
+    // streamed directly to cerr.
+    // Use with caution as this function isn't MT_SAFE.
+    static string warnMoreStandalone() VL_EXCLUDES(s().m_mutex) VL_MT_UNSAFE {
+        const VerilatedLockGuard guard{s().m_mutex};
+        return s().warnMore();
+    }
+    // This function marks place in error message from which point message
+    // should be printed after information about error code.
+    // This post-processing is done in v3errorEnd function.
```
Comment (wsnyder) [inline src/V3Error.h:469]:
```suggestion
    // should be printed after information on the error code.
    // The post-processing is done in v3errorEnd.
```

File: src/V3Error.h [hunk 16]
```diff
@@ -323,36 +476,65 @@ class V3Error final {
...
     static void vlAbort();
     // static, but often overridden in classes.
-    static void v3errorEnd(std::ostringstream& sstr, const string& extra = "");
+    static void v3errorEnd(std::ostringstream& sstr, const string& extra = "")
+        VL_MT_SAFE_EXCLUDES(s().m_mutex) VL_MT_SAFE {
+        const VerilatedLockGuard guard{s().m_mutex};
+        s().v3errorEnd(sstr, extra);
+    }
+    // We can't call 's().v3errorEnd' directly in 'v3ErrorEnd'/'v3errorEndFatal',
+    // due to bug in gcc causing internal error when backtrace is printed.
```
Comment (wsnyder) [inline src/V3Error.h:495]: Please indicate the version of gcc with that bug (and capitalize GCC).

File: src/V3LinkLevel.cpp (hunk 13 above)
Comment (kamilrakoczy) [reply src/V3LinkLevel.cpp:71]: This lock is not for ``warnContextSecondary`` but for ``warnMore`` (it is passed as argument to lambda as its value is the same for all modules, but we need to have V3Error lock to make sure we always get correct value)

**Comment by wsnyder** [comment]
Adding a lot of code for warnMore I'm not sure is worth it.  We previously had:

```
%Warning-MULTITOP: t/t_multitop_sig.v:15:8: Multiple top level modules
                                          : ... Suggest see manual; fix the duplicates, or use --top-module to select top.
                   ... For warning description see https://verilator.org/warn/MULTITOP?v=latest
                   ... Use "/* verilator lint_off MULTITOP */" and lint_on around source to disable this message.
                                          : ... Top module 'a'
    7 | module a(in, out);
      |        ^
```

Maybe we just make all warnMore's a fixed indent of 8

```
%Warning-MULTITOP: t/t_multitop_sig.v:15:8: Multiple top level modules
        : ... Suggest see manual; fix the duplicates, or use --top-module to select top.
        : ... For warning description see https://verilator.org/warn/MULTITOP?v=latest
        : ... Use "/* verilator lint_off MULTITOP */" and lint_on around source to disable this message.
        : ... Top module 'a'
    7 | module a(in, out);
      |        ^
```

The upside is easier code, and also less wide lines for long messages.  Thoughts?   If this is agreed I'll make the change outside of this, as I want to see how the messages all look.

**Comment by kamilrakoczy** [comment]
@wsnyder Is there decision how we want to proceed with this?

**Comment by wsnyder** [comment]
I tried the new error alignment I suggested and didn't like it.  I don't think you need to wait for that change.

Please update from master, make sure this runs on Ubuntu 22.04/20.04/18.04 (more than default pull github action), then you may squash merge, thanks for being patient on getting this ready.

**Comment by kamilrakoczy** [comment]
@wsnyder I've rebased from master and squashed commits. I've also checked that push CI jobs (https://github.com/antmicro/verilator-1/actions/runs/4134275089) and Ubuntu 18.04 (locally in docker, as we don't have CI job for this version) also passes.

I can't merge, as I don't have permissions to do this.
